corehq/apps/change_feed/consumer/feed.py

```diff
@@ -201,3 +206,86 @@ def change_from_kafka_message(message):
 
 def change_meta_from_kafka_message(message):
     return ChangeMeta.wrap(json.loads(message))
+
+
+def change_from_postgres_model(change):
+    change_meta = ChangeMeta(
+        document_id=change.get_id,
+        data_source_type='sql',
+        data_source_name='form-sql',
+        document_type=change.doc_type,
+        document_subtype=change.xmlns,
+        domain=change.domain,
+        is_deletion=change.is_deleted,
+        publish_timestamp=change.server_modified_on,
+    )
+    return Change(
+        id=change_meta.document_id,
+        sequence_id=change.update_sequence_id,
+        document=change.to_json(include_attachments=True),
+        deleted=change_meta.is_deletion,
+        metadata=change_meta,
+    )
+
+
+@lru_cache()
+def _model_class_from_name(model_name):
+    return to_function(model_name)
+
+
+class PostgresChangeFeed(ChangeFeed):
+    def __init__(self, pillow_id):
+        self.pillow_id = pillow_id
+
+    def iter_changes(self, since, forever: bool):
+        number_changes = -1
+
+        while number_changes != 0 or forever:
+            number_changes = 0
+            with transaction.atomic():
```

Comment:
It looks like all change processing is done inside this transaction. Is that true? Would you expect performance problems if some changes take a long time? Is that a concern with our existing change processors?

Reply:
In the existing change processors, only two write to the same SQL database that this model resides on (`mark_latest_submission` and `mark_latest_synclog`) so these transactions are not changing much data and are only acting as a lock on the row (using `select_for_update` inside the transaction). 

In any case, long running transactions are not ideal and I would expect to configure `batch_size` to limit the amount spent in the transaction.